### PR TITLE
Improve MultiAlters error handling with Alter Primary Key, Alter Index, Alter Default, Alter AutoIncrement

### PR DESCRIPTION
**Pull request by VinaiRachakonda**
This pr does the following

1. Improves MultiAlter DDL analyzer by checking for failure modes in adding/dropping pks, adding/dropping indexes, adding/dropping defaults
2. Pushes the table resolution of AlterPk, AlterIndex, AlterDefault,Alter AutoIncrement at execution time. This is due to the fact that the node should know whether columns were added or dropped in previous steps of a MultiAlter DDL.
3. Adds a small bug fix with how columns were being dropped in the validate_create_table 
4. This pr does not include changes to multialters with ADD/DROP constraints due to additional complexity.

### CODE REVIEW

File: sql/plan/alter_auto_increment.go [hunk 1]
```diff
@@ -21,20 +21,34 @@ import (
...
+		Table:   table,
+		autoVal: autoVal,
 	}
 }
 
 // Execute inserts the rows in the database.
 func (p *AlterAutoIncrement) Execute(ctx *sql.Context) error {
-	insertable, err := GetInsertable(p.UnaryNode.Child)
+	// Grab the table fresh from the database.
+	table, ok, err := p.ddlNode.Database().GetTableInsensitive(ctx, getTableName(p.Table))
```
Comment (VinaiRachakonda) [inline sql/plan/alter_auto_increment.go:40]: I noticed there is repeated code here. I will address this post review.

File: sql/analyzer/validate_create_table.go [hunk 2]
```diff
@@ -145,7 +180,9 @@ func validateAddColumn(initialSch sql.Schema, schema sql.Schema, ac *plan.AddCol
 	}
 
 	// None of the checks we do concern ordering, so we don't need to worry about it here
-	newSch := append(table.Schema(), ac.Column())
+	newCol := copyColumn(ac.Column())
```
Comment (zachmu) [inline sql/analyzer/validate_create_table.go:183]: Would add Column.Copy() in the sql package, generally useful for things like this

File: sql/analyzer/validate_create_table.go [hunk 3]
```diff
@@ -207,6 +244,57 @@ func validateColumnNotUsedInCheckConstraint(columnName string, checks sql.CheckC
 	return nil
 }
 
+// validateAlterIndex validates the specified column can have an index added, dropped, or renamed.
```
Comment (zachmu) [inline sql/analyzer/validate_create_table.go:247]: Mention what the return value is

File: sql/analyzer/validate_create_table.go [hunk 4]
```diff
@@ -207,6 +244,57 @@ func validateColumnNotUsedInCheckConstraint(columnName string, checks sql.CheckC
...
+		badColName, ok := schContainsAllIndexColumns(ai.Columns, sch, tableName)
+		if !ok {
+			return nil, sql.ErrKeyColumnDoesNotExist.New(badColName)
+		}
+
+		return append(indexes, ai.IndexName), nil
+	case plan.IndexAction_Drop:
+		for _, idx := range indexes {
+			if strings.EqualFold(idx, ai.IndexName) {
+				return indexes, nil
```
Comment (zachmu) [inline sql/analyzer/validate_create_table.go:262]: This doesn't seem right, you're not removing an index here. Should have a test.

File: sql/analyzer/validate_create_table.go [hunk 5]
```diff
@@ -207,6 +244,57 @@ func validateColumnNotUsedInCheckConstraint(columnName string, checks sql.CheckC
...
+		// Simulate the rename by deleting the old name and adding the new one.
+		return append(append(indexes[:savedIdx], indexes[savedIdx+1:]...), ai.IndexName), nil
+	}
+
+	return indexes, nil
+}
+
+// schContainsAllIndexColumns takes in a set of IndexColumns and returns false, along with the offending column name, if
+// an index Column is not in an index.
+func schContainsAllIndexColumns(cols []sql.IndexColumn, sch sql.Schema, tableName string) (string, bool) {
```
Comment (zachmu) [inline sql/analyzer/validate_create_table.go:288]: missingIndexCol

**Review by zachmu — APPROVED** [review]
Generally quite good, just a few small changes

**Pull request merged**